File: vc-bdd-scenarios/features/step_definitions/age-gate.steps.ts
import { Then, When } from "@cucumber/cucumber";
import { Ensure, equals } from "@serenity-js/assertions";
import { actorCalled } from "@serenity-js/core";

import {
  AgeGateScenarioOutcome,
  RunTheBirthCredentialAgeGateHappyPath,
} from "../support/tasks.js";

const engineer = () => actorCalled("Engineer");

When("the engineer runs the birth credential age-gate happy path", async () => {
  await engineer().attemptsTo(RunTheBirthCredentialAgeGateHappyPath());
});

Then("the scenario should be approved", async () => {
  await engineer().attemptsTo(
    Ensure.that(AgeGateScenarioOutcome.approved(), equals(true)),
  );
});

Then("the scenario should issue {int} credential", async (expectedCount: number) => {
  await engineer().attemptsTo(
    Ensure.that(
      AgeGateScenarioOutcome.issuedCredentialCount(),
      equals(BigInt(expectedCount)),
    ),
  );
});

Then(
  "the scenario should verify {int} presentation",
  async (expectedCount: number) => {
    await engineer().attemptsTo(
      Ensure.that(
        AgeGateScenarioOutcome.verifiedPresentationCount(),
        equals(BigInt(expectedCount)),
      ),
    );
  },
);

Then(
  "the scenario should consume {int} access capability",
  async (expectedCount: number) => {
    await engineer().attemptsTo(
      Ensure.that(
        AgeGateScenarioOutcome.consumedCapabilityCount(),
        equals(BigInt(expectedCount)),
      ),
    );
  },
);

File: serenity-vc-scenarios/src/test/java/io/iohk/midnight/vc/serenity/steps/AgeGateSteps.java
package io.iohk.midnight.vc.serenity.steps;

import io.iohk.midnight.vc.serenity.ScenarioBridge;
import io.iohk.midnight.vc.serenity.SerenityScenarioResult;
import net.serenitybdd.annotations.Step;
import net.serenitybdd.core.steps.UIInteractionSteps;
import org.assertj.core.api.Assertions;

public class AgeGateSteps extends UIInteractionSteps {
  private final ScenarioBridge bridge = new ScenarioBridge();
  private SerenityScenarioResult result;

  @Step("Run the birth credential age-gate happy-path scenario")
  public void runAgeGateHappyPath() throws Exception {
    bridge.assertBuildPrerequisitesPresent();
    result = bridge.runNodeScenario("serenity-vc-scenarios/scripts/age-gate-happy-path.mjs");
  }

  @Step("The scenario should be approved")
  public void scenarioShouldBeApproved() {
    Assertions.assertThat(result.approved()).isTrue();
    Assertions.assertThat(result.claimDecision()).isEqualTo("approved");
    Assertions.assertThat(result.lastVerifiedCredentialRoot())
        .isEqualTo(result.expectedCredentialRoot());
  }

  @Step("The scenario should issue {0} credential")
  public void scenarioShouldIssueCredentials(long expectedCount) {
    Assertions.assertThat(result.issuedCredentialCount()).isEqualTo(expectedCount);
  }

  @Step("The scenario should verify {0} presentation")
  public void scenarioShouldVerifyPresentations(long expectedCount) {
    Assertions.assertThat(result.verifiedPresentationCount()).isEqualTo(expectedCount);
  }

  @Step("The scenario should consume {0} access capability")
  public void scenarioShouldConsumeCapabilities(long expectedCount) {
    Assertions.assertThat(result.consumedAccessCapabilityCount()).isEqualTo(expectedCount);
  }
}
